managed/services/management/agent_test.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package management
+
+import (
+	"context"
+	"testing"
+
+	"github.com/AlekSi/pointer"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/reform.v1"
+	"gopkg.in/reform.v1/dialects/postgresql"
+
+	"github.com/percona/pmm/api/managementpb"
+	"github.com/percona/pmm/managed/models"
+	"github.com/percona/pmm/managed/utils/logger"
+	"github.com/percona/pmm/managed/utils/testdb"
+	"github.com/percona/pmm/managed/utils/tests"
+)
+
+func TestAgentService(t *testing.T) {
+	setup := func(t *testing.T) (context.Context, *AgentService, func(t *testing.T), *mockPrometheusService) { //nolint:unparam
+		t.Helper()
+
+		ctx := logger.Set(context.Background(), t.Name())
+		uuid.SetRand(&tests.IDReader{})
+
+		sqlDB := testdb.Open(t, models.SetupFixtures, nil)
+		db := reform.NewDB(sqlDB, postgresql.Dialect, reform.NewPrintfLogger(t.Logf))
+
+		vmdb := &mockPrometheusService{}
+		vmdb.Test(t)
+
+		state := &mockAgentsStateUpdater{}
+		state.Test(t)
+
+		ar := &mockAgentsRegistry{}
+		ar.Test(t)
+
+		teardown := func(t *testing.T) {
+			uuid.SetRand(nil)
+
+			require.NoError(t, sqlDB.Close())
+			vmdb.AssertExpectations(t)
+			state.AssertExpectations(t)
+			ar.AssertExpectations(t)
+		}
+		s := NewAgentService(db, ar, state, vmdb)
+
+		return ctx, s, teardown, vmdb
+	}
+
+	t.Run("List", func(t *testing.T) {
+		t.Run("Basic", func(t *testing.T) {
```

Comment:
Why do we need this nested func?

Reply:
It seems it's an established patter if we look [here](https://github.com/percona/pmm/blob/main/managed/services/management/service_test.go#L68-L78) for example. That makes sense to me since we are testing different scenarios for the same endpoint. 

Apart from that, it reads well, like a hierarchy on a very voluminous test, which is helpful when navigating through a long output.

That said, we can unwrap of course. Pls let me know what you prefer.